vendor/bat-native-ledger/src/bat/ledger/internal/report/report.cc

```diff
@@ -0,0 +1,168 @@
+/* Copyright (c) 2020 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include <algorithm>
+#include <utility>
+#include <vector>
+#include <iostream>
+
+#include "base/strings/string_split.h"
+#include "bat/ledger/internal/common/bind_util.h"
+#include "bat/ledger/internal/ledger_impl.h"
+#include "bat/ledger/internal/report/report.h"
+
+using std::placeholders::_1;
+using std::placeholders::_2;
+
+namespace braveledger_report {
+
+Report::Report(bat_ledger::LedgerImpl* ledger):
+    ledger_(ledger) {
+  DCHECK(ledger_);
+}
+
+Report::~Report() = default;
+
+void Report::GetMonthly(
+      const ledger::ActivityMonth month,
+      const int year,
+      ledger::GetMonthlyReportCallback callback) {
+  auto balance_callback = std::bind(&Report::OnBalance,
+      this,
+      _1,
+      _2,
+      month,
+      year,
+      callback);
+
+  ledger_->GetBalanceReport(month, year, balance_callback);
+}
+
+void Report::OnBalance(
+    const ledger::Result result,
+    ledger::BalanceReportInfoPtr balance_report,
+    const ledger::ActivityMonth month,
+    const uint32_t year,
+    ledger::GetMonthlyReportCallback callback) {
+  if (result != ledger::Result::LEDGER_OK || !balance_report) {
+    BLOG(ledger_, ledger::LogLevel::LOG_ERROR)
+        << "Could not get balance report";
+    callback(result, nullptr);
+    return;
+  }
+
+  auto monthly_report = ledger::MonthlyReportInfo::New();
+  monthly_report->balance = std::move(balance_report);
+
+  const std::string monthly_report_string =
+      braveledger_bind_util::FromMonthlyReportToString(
```

Comment:
It seems like we do this JSON parse/serialize step a few times. Is there any straightforward way that we can eliminate those round-trips?

Reply:
this is the only way that we found so far that enables us sending mojo structs through std::bind where we have nested mojo objects